src/charm.py

```diff
@@ -152,7 +159,74 @@ def temp_maintenance_status(self, message: str):
         yield
         self.unit.status = previous_status
 
-    def _on_collect_status(self, event: CollectStatusEvent):
+    def _on_authorize_charm_action(self, event: ActionEvent):
+        """Authorize the charm to interact with Vault."""
+        if not self.unit.is_leader():
+            event.fail("This action can only be run by the leader unit")
+            return
+        logger.info("Authorizing the charm to interact with Vault")
+        if not self._api_address:
+            event.fail("API address is not available.")
+            return
+        if not self.tls.tls_file_available_in_charm(File.CA):
+            event.fail("CA certificate is not available in the charm. Something is wrong.")
+            return
+        token = event.params["token"]
+        vault = self._get_vault_client()
+        vault.authenticate(Token(token))
+        try:
+            vault.enable_audit_device(device_type=AuditDeviceType.FILE, path="stdout")
+            vault.enable_approle_auth_method()
+            vault.configure_policy(policy_name=CHARM_POLICY_NAME, policy_path=CHARM_POLICY_PATH)
+            role_id = vault.configure_approle(
+                role_name="charm",
+                policies=[CHARM_POLICY_NAME, VAULT_DEFAULT_POLICY_NAME],
+            )
+            vault_secret_id = vault.generate_role_secret_id(name="charm")
+            self._create_approle_secret(role_id, vault_secret_id)
+        except VaultError as e:
+            logger.exception("Vault returned an error while authorizing the charm")
+            event.fail(f"Vault returned an error while authorizing the charm: {str(e)}")
+            return
+        self._configure(event)
+
+    def _create_approle_secret(self, role_id: str, secret_id: str):
+        secret = self.app.add_secret(
+            {"role-id": role_id, "secret-id": secret_id},
+            label=VAULT_CHARM_APPROLE_SECRET_LABEL,
+        )
+        assert (
+            secret and secret.id
+        ), f"Unexpected error while attempting to create secret {VAULT_CHARM_APPROLE_SECRET_LABEL}"
+        return secret
+
+    def _get_vault_client(self) -> Vault:
+        assert self._api_address
+        assert self.tls.tls_file_available_in_charm(File.CA)
```

Comment:
This could crash the charm, but I have mixed feelings on this. I'd like things to break when we have obvious mistakes, and these things should be verified in every case before this. I don't like returning `None` because that just hides the error and makes it much more difficult to debug.

Would be interested in hearing others' perspective here.

Reply:
Discussion continued in https://github.com/canonical/vault-operator/pull/62#discussion_r1531972101